cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go

```diff
@@ -0,0 +1,253 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package phases
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/options"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow"
+	cmdutil "k8s.io/kubernetes/cmd/kubeadm/app/cmd/util"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/controlplane"
+	etcdphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/etcd"
+	markcontrolplanephase "k8s.io/kubernetes/cmd/kubeadm/app/phases/markcontrolplane"
+	uploadconfigphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/uploadconfig"
+	"k8s.io/kubernetes/pkg/util/normalizer"
+)
+
+var (
+	controlPlaneJoinExample = normalizer.Examples(`
+		# Joins a machine as a control plane instance
+		kubeadm join phase control-plane-join all
+		`)
+
+	controlPlanePhaseProperties = map[string]struct {
+		name  string
+		short string
+	}{
+		kubeadmconstants.KubeAPIServer: {
+			name:  "apiserver",
+			short: getPhaseDescription(kubeadmconstants.KubeAPIServer),
+		},
+		kubeadmconstants.KubeControllerManager: {
+			name:  "controller-manager",
+			short: getPhaseDescription(kubeadmconstants.KubeControllerManager),
+		},
+		kubeadmconstants.KubeScheduler: {
+			name:  "scheduler",
+			short: getPhaseDescription(kubeadmconstants.KubeScheduler),
+		},
+	}
+)
+
+type controlPlaneJoinData interface {
+	Cfg() *kubeadmapi.JoinConfiguration
+	ClientSetFromFile(string) (*clientset.Clientset, error)
+	InitCfg() (*kubeadmapi.InitConfiguration, error)
+}
+
+func getPhaseDescription(component string) string {
+	return fmt.Sprintf("Generates the %s static Pod manifest", component)
+}
+
+func getControlPlaneJoinPhaseFlags() []string {
+	return []string{
+		options.APIServerAdvertiseAddress,
+		options.APIServerBindPort,
+		options.CfgPath,
+		options.ControlPlane,
```

Comment:
Ideally, we should not need this one. But in the current form of the PR, I don't think, that we can operate without either this one being specified on the command line or a config file.

Reply:
I think we actually need it in case no config is provided, the `--experimental-control-plane` needs to be provided so we know how to branch in order to join a master and not a worker.

Reply:
On regular join `--experimental-control-plane` will be dealt for in `newJoinData` prior to calling the actual phase (which is the case we don't know if we should join a worker or control plane node).
If the phase is invoked directly via the command line interface (`kubeadm join phase control-plane-join ...`) then we already know, that the user wants to join a control plane and do that accordingly.

Reply:
I agree with this. However the rest of the phases have this flag IIRC. In my opinion we could go this same way on this last one, and refactor all phases afterwards when we perform the UX audit. WDYT?